src/line_iterator.rs

```diff
@@ -65,11 +65,42 @@ impl<R: BufRead> Iterator for LineIterator<R> {
                     } else {
                         // It's the new format! Drop next line.
                         // (Hmm, is there a way to do this without an allocation?)
-                        let mut devnull = String::new();
-                        if let Err(e) = self.reader.read_line(&mut devnull) {
+                        // let mut devnull = String::new();
+                        // if let Err(e) = self.reader.read_line(&mut devnull) {
+                        if let Err(e) = skip_one_line(&mut self.reader) {
                             warn!("Could not read line from reader: {:?}", e);
                             return None;
                         }
+
+                        fn skip_one_line<R: BufRead>(r: &mut R) -> std::io::Result<()> {
+                            loop {
+                                // get inner buffer
+                                let buf = r.fill_buf()?;
+                                // no b'\n' found until EOF
+                                if buf.is_empty() {
+                                    break;
+                                }
+                                // save buf.len() to avoid hold `&mut R` caused by BufRead::fill_buf(&mut self)
+                                let buf_len = buf.len();
+                                // prepare the argument which BufRead::consume(&mut self, amt: usize) need
+                                let mut amt = buf_len;
+                                // search b'\n' in the buffer from left to right
+                                for (&ch, index) in buf.iter().zip(1usize..) {
+                                    if ch == b'\n' {
+                                        amt = index;
+                                        break;
+                                    }
+                                }
+                                // inform BufRead how many bytes should be consumed
+                                r.consume(amt);
+                                // no more loops are needed if the byte b'\n' has been found in the buffer
+                                if amt != buf_len {
+                                    break;
+                                }
+                            }
+                            Ok(())
```

Comment:
I fell into a bit of a rabbit hole with the implementation. At first, it can be simplified by using `Iterator::position`:
```rust
loop {
    let buf = r.fill_buf()?;

    if buf.is_empty() {
        return Ok(());
    }

    if let Some(offset) = buf.iter().position(|&c| c == b'\n') {
        r.consume(offset + 1);
        return Ok(());
    }

    let buf_len = buf.len();
    r.consume(buf_len);
}
```

However, this still feels too complicated - we don't really care about buffers and filling and consuming them! We really just want to read from the reader until there is a newline. So, we look at the bytes of the reader and `find` the first newline character or error. By `find`ing up to this point, the reader is advanced exactly as far as we want. Finally, we discard whatever the result of the search was into `()`:
```rust
<&mut R as std::io::Read>::bytes(r)
    .find(|b| matches!(b, Ok(b'\n') | Err(_)))
    .transpose()
    .map(|_newline_char| ())
```
This version is far more concise, but a bit harder to understand because we manipulate `Option<Result<...>>` and need to avoid `Deref` on `r`. However, it works for any type that is `Read`, not only `BufRead` and is closer to what we actually want to do.

I like the second version more, but am fine with either of the two. @dbrgn what do you think?

Reply:
Your solution is much better! Now we can do the same thing in only 3 lines of codes, not in 30 or more. You are genius!